ecs-agent/acs/session/session_test.go

```diff
@@ -0,0 +1,1328 @@
+	mockWsClient.EXPECT().Serve(gomock.Any()).Return(inactiveInstanceError).AnyTimes()
+
+	acsSession := session{
+		containerInstanceARN:           testconst.ContainerInstanceARN,
+		discoverEndpointClient:         discoverEndpointClient,
+		inactiveInstanceCB:             noopFunc,
+		clientFactory:                  mockClientFactory,
+		heartbeatTimeout:               20 * time.Millisecond,
+		heartbeatJitter:                10 * time.Millisecond,
+		inactiveInstanceReconnectDelay: 1 * time.Hour,
+		backoff: retry.NewExponentialBackoff(connectionBackoffMin, connectionBackoffMax,
+			connectionBackoffJitter, connectionBackoffMultiplier),
+	}
+
+	err := acsSession.Start(ctx)
+	assert.Equal(t, context.DeadlineExceeded, err)
+}
+
+// TestSessionReconnectsOnDiscoverPollEndpointError tests that the Session retries to establish the connection with ACS
+// on DiscoverPollEndpoint errors.
+func TestSessionReconnectsOnDiscoverPollEndpointError(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	discoverEndpointClient := mock_api.NewMockECSDiscoverEndpointSDK(ctrl)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	mockWsClient := mock_wsclient.NewMockClientServer(ctrl)
+	mockClientFactory := mock_wsclient.NewMockClientFactory(ctrl)
+	mockClientFactory.EXPECT().
+		New(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+		Return(mockWsClient).AnyTimes()
+	mockWsClient.EXPECT().SetAnyRequestHandler(gomock.Any()).AnyTimes()
+	mockWsClient.EXPECT().AddRequestHandler(gomock.Any()).AnyTimes()
+	mockWsClient.EXPECT().Serve(gomock.Any()).AnyTimes()
+	mockWsClient.EXPECT().WriteCloseMessage().Return(nil).AnyTimes()
+	mockWsClient.EXPECT().Close().Return(nil).AnyTimes()
+	mockWsClient.EXPECT().Connect(gomock.Any(), gomock.Any(), gomock.Any()).Do(func(interface{},
+		interface{}, interface{}) {
+		// Serve() cancels the context
+		cancel()
+	}).Return(time.NewTimer(wsclient.DisconnectTimeout), nil).MinTimes(1)
+
+	gomock.InOrder(
+		// DiscoverPollEndpoint returns an error on its first invocation.
+		discoverEndpointClient.EXPECT().DiscoverPollEndpoint(gomock.Any()).Return("", fmt.Errorf("oops")),
+		// Second invocation returns a success.
+		discoverEndpointClient.EXPECT().DiscoverPollEndpoint(gomock.Any()).Return(acsURL, nil),
+	)
+	acsSession := session{
+		containerInstanceARN:   testconst.ContainerInstanceARN,
+		discoverEndpointClient: discoverEndpointClient,
+		inactiveInstanceCB:     noopFunc,
+		clientFactory:          mockClientFactory,
+		heartbeatTimeout:       20 * time.Millisecond,
+		heartbeatJitter:        10 * time.Millisecond,
+		disconnectTimeout:      30 * time.Millisecond,
+		disconnectJitter:       10 * time.Millisecond,
+		backoff: retry.NewExponentialBackoff(connectionBackoffMin, connectionBackoffMax,
+			connectionBackoffJitter, connectionBackoffMultiplier),
+	}
+
+	start := time.Now()
+	err := acsSession.Start(ctx)
+	assert.Equal(t, context.Canceled, err)
```

Comment:
bonus: maybe we could assert something to make sure no more ACS data got set? Like `sendCredentials` is true on both reconnects (set in `session.go` L255)

Reply:
I don't think we need to check `sendCredentials` is as expected for this test (or every test) as this test is not testing that and we have a specific test for that (i.e., `TestSessionCorrectlySetsSendCredentials`).

Reply:
that is true about `sendCredentials` being out of scopr, but should we assert that the first failed `DiscoverPollEndpoint()` in this test results in no connection to ACS? Or, that the second call results in a stable connection to ACS?

Reply:
We implicitly know that the second call results in a stable connection (whereas the first call didn't) since expectation of the .Connect method being called on L707-711 is met and the context is then canceled (which is asserted on L734).

Reply:
ahh, thank you for the explanation - It may be worth adding this as a comment in the test.

Reply:
No worries and sure, will add a small note for that.